### ISSUE TITLE
rfe: grouping of arguments

### ISSUE THREAD

**Issue by ibotty** [pcapriotti/optparse-applicative#187]
Say, I have a parser that has some flags that always belong together (say, source and destination), and I want to parse many of them. I can (and do) use `many` (or `some`) and `zip`, to parse them. The problem is, that it's not possible to specify default arguments that way.

I hope, that's still in optparse-applicative's scope, because it technically needs monadic parsing. The interface could still be applicative though, so I have some hope ;).

If I should clarify some parts, please tell me.

**Comment by HuwCampbell** [comment]
Please do clarify. Can you show me the code, some current inputs and outputs, and what you would like to see different in the output?

**Comment by ibotty** [comment]
Well, I can't show code because it is not possible now afaict. What I would like is the following.

An invocation of `cmd -s s1 -s s2 -d d2` should be parsed into `[("s1",Nothing),("s2", Just "d2")]`. Does that make it more clear?

**Comment by HuwCampbell** [comment]
I've been having a think about this. Currently, when running `searchParser` on `BindP`, we run `evalParser` to see if the complete parser succeeded, and if not, return `mzero`.

We could potentially change this to:
```haskell
searchParser f (BindP p k) = do
  p' <- searchParser f p
  case evalParser p' of
    Nothing -> pure $ BindP p' k
    Just aa -> pure $ k aa
```

Which would allow pairs to be written as
```haskell
let x = some $ (,) <$> strOption (short 's') <*> (strOption (short 'd'))
```

But, I think we'd have to deal with a lot of edge cases creeping in. If one of these was optional, it would jump to the next level greedily, and what may seem to be sensible pairing to the user may not be to the parser.

Another option for your problem, which I would suggest for now, is to use the ReadM to parse two colon separated values. Then your input could be something like
```haskell 
many $ argument pairReader (metavar "PAIR")
  where
    pairReader :: ReadM (FilePath, FilePath)
```
And your input could be
```
cmd s1:d1 s2:d2
```

**Comment by HuwCampbell** [comment]
I think I can solve the problem of dropping to the next level too quickly as well which is nice.

**Comment by HuwCampbell** [comment]
See https://github.com/pcapriotti/optparse-applicative/commit/e7b9d27cb15e5745e93174d213e75c62ffd5e5e2
For a tentative solution. The tests should show how it can be used.

**Comment by ibotty** [comment]
That looks great. I will play around with it this week.For file names separating by : is not so nice because : is a valid character in file names.pcapriotti/optparse-applicative wrote:See e7b9d27
For a tentative solution. The tests should show how it can be used.

**Comment by HuwCampbell** [comment]
Fixed in #191
